concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -150,90 +251,186 @@ exportDatabaseV3 ::
-                -- write back the index file without the last line (corresponding to the chunk deleted above)
-                writeFile (indexFile ++ "~") . unlines . init . lines $ indexContents
-                renameFile (indexFile ++ "~") indexFile
-                return (lastChunkGenesisIndex, lastChunkFirstBlock)
+                blockIndex <- case AP.parseOnly parseBlockIndexFile (T.pack indexContents) of
+                    Left err -> fail $ "An error occurred while parsing '" <> indexFile <> "': " <> err
+                    -- normalize the block index, s.t. superfluous genesis hash lines are removed.
+                    Right bi -> do
+                        return $ normalizeBlockIndex bi
+                case blockIndex of
+                    -- we export blocks starting from height 1 because genesis blocks need not be exported.
+                    Empty -> return (0, 1, Empty, Nothing)
+                    -- the following is safe due to normalization of the blockindex.
+                    blockIndexInit :|> (gd, chunks) -> case chunks of
+                        Empty -> fail $ "A section for genesis hash '" <> show gd <> "' in block index file has no entries."
+                        initChunks :|> BlockIndexChunkInfo{..} -> do
+                            -- the filename of the most recently exported chunk
+                            let lastChunkFile = outDir </> T.unpack filename
+                            -- if we overwrite files, back up the most recently exported
+                            -- chunk file so we can restore it if something goes wrong.
+                            when overwriteExistingChunks $ renameFile lastChunkFile $ lastChunkFile <> "~"
+                            -- evict the last chunk from the block index since we want to
+                            -- export blocks starting from the genesis index and the height
+                            -- of the first block contained within it
+                            return
+                                ( genesisIndex,
+                                  blockHeightFirst,
+                                  blockIndexInit :|> (gd, initChunks),
+                                  Just lastChunkFile
+                                )
             else do
                 createDirectoryIfMissing True outDir
-                -- we export blocks starting from height 1 because genesis blocks need not be exported
-                return (0, 1)
-    bracket
-        (openFile indexFile AppendMode)
-        hClose
-        (\indexHdl -> exportSections dbDir outDir indexHdl chunkSize genIndex startHeight)
-
--- |Export a database section for each genesis index.
+                -- we export blocks starting from height 1 because genesis blocks need not be exported.
+                return (0, 1, Empty, Nothing)
+
+    let lastChunkFileName = fromMaybe "" lastChunkFilenameM
```

Comment:
Why is this sensible? The empty string is generally not a valid file name. At the very least mention why this is sensible.